src/index.ts

```diff
@@ -0,0 +1,354 @@
+import {Activity, ActivityTypes, BotAdapter, ConversationAccount, ConversationReference, ChannelAccount, ResourceResponse, TurnContext, WebRequest, WebResponse} from "botbuilder";
+import {AxiosInstance} from "axios";
+import {ITyntecMoMessage, ITyntecWhatsAppMessageRequest} from "./tyntec/messages";
+import {composeTyntecRequestConfig, composeTyntecSendWhatsAppMessageRequestConfig, parseTyntecSendWhatsAppMessageResponse} from "./tyntec/axios";
+
+export interface ITyntecWhatsAppAdapterSettings {
+    axiosInstance: AxiosInstance;
+    maxBodySize?: number;
+    tyntecApikey: string;
+}
+
+export class TyntecWhatsAppAdapter extends BotAdapter {
+    public axiosInstance: AxiosInstance;
+    public maxBodySize = 1024;
+    public tyntecApikey: string;
+
+    constructor(settings: ITyntecWhatsAppAdapterSettings) {
+        super();
+
+        this.axiosInstance = settings.axiosInstance;
+        if (settings.maxBodySize !== undefined) {
+            this.maxBodySize = settings.maxBodySize;
+        }
+        this.tyntecApikey = settings.tyntecApikey;
+    }
+
+    async continueConversation(reference: Partial<ConversationReference>, logic: (revocableContext: TurnContext) => Promise<void>): Promise<void> {
+        throw Error("Operation continueConversation not supported.");
+    }
+
+    async deleteActivity(context: TurnContext, reference: Partial<ConversationReference>): Promise<void> {
+        throw Error("Operation deleteActivity not supported.");
+    }
+
+    async processActivity(req: WebRequest, res: WebResponse, logic: (context: TurnContext) => Promise<any>): Promise<void> {
+        try {
+            const requestBody = await new Promise((resolve: (value: ITyntecMoMessage) => void, reject: (reason?: any) => void) => {
+                if (req.body !== undefined) {
+                    return resolve(req.body);
+                }
+
+                let requestJson = '';
+                req.on!('data', (chunk: string) => {
```

Comment:
It's a non-null assertion operator: https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#non-null-assertion-operator-postfix-

Here, I am telling to TypeScript that, although it believes that `on` may be undefined (as per the [`WebRequest` declaration](https://github.com/microsoft/botbuilder-js/blob/4.14/libraries/botbuilder/src/interfaces/webRequest.ts#L50)), **`on` is not undefined nor `null` at this point in the code**. If I remove the exclamation mark, TypeScript will not compile the code and I would have to add an explicit check (`if (!req.on)`) which is pointless here since I believe that the type declaration of `WebRequest` is wrong and that `on` is always defined.

Reply:
I could not resist and had to find out why they made `on` optional. [It seems that in some frameworks `on` really is optional](https://github.com/microsoft/botbuilder-js/issues/1589) (although [Bot Framework does not seem to support these frameworks](https://docs.microsoft.com/en-us/javascript/api/botbuilder/webrequest?view=botbuilder-ts-latest)). But in such cases `req.body` is not `null` so it should be OK in my code.

Reply:
Ahaaa. I see, I know about this operator and I use it frequently but I didn't see it used for a handler like this, I thought as well that it should be defined by default so it looked a bit weird to me. Good, thanks for clarification! Other than that I didn't find anything unusual in the code 👍

Reply:
Do you mean that you approve the PR? :star_struck:

Reply:
yes :)